### ISSUE TITLE
Official request for linux support

### ISSUE THREAD

**Issue by ikt** [pdaddyo/soundbounce#114]
Didn't see any other requests for it so here I am! Would love a linux version since I use Ubuntu 99% of my day.

This project looks fantastic, can't wait to see it take off :)

Edit: for the bounty I'm looking for a 64bit Ubuntu 14.04 compatible .deb that installs everything nicely
##

**Comment by pdaddyo** [comment]
Our adjusted node-spotify library will need recompiling for linux node-webkit using nw-gyp, and then the existing app.nw file should load on linux using node-webkit.  I don't know much more than that I'm afraid, good luck!

**Comment by Beamboom** [comment]
+1 for Linux support. I'm a 100% Linux user, former Soundrop user who are desperately looking for a replacement.

**Comment by earthling** [comment]
I think I've got everything built and running on Linux, but I need an 'app.key' - is each Soundbounce user supposed to get their own app key? or is it included in the soundbouce distribution? It seems to have been intentionally excluded from source control. Help?

**Comment by earthling** [comment]
I extracted the app.key from the osx distribution and it's working on Linux! 
I don't really know anything about nodejs et. al., so aplogies if this is hamfisted, but here's what I did:
1. You cannot use node-gyp and node to build and run this application. You need to use [node-webkit (0.8.6)](http://dl.node-webkit.org/v0.8.6/node-webkit-v0.8.6-linux-x64.tar.gz) and [nw-gyp](https://github.com/rogerwang/node-webkit/wiki/Build-native-modules-with-nw-gyp) - which is webkit bundled with nodejs.
   * Change the linker arguments in node-spotify-master/binding.gyp to:
```
    "link_settings" : { 
      "libraries": ["-lasound -lspotify"]
    },  
```
   * To rebuild node-spotify-master, you must use nw-gyp and set --target to 0.8.6 (the node bindings for libspotify will not compile against the latest release).
```
  $ nw-gyp configure --target=0.8.6
  $ nw-gyp build --target=0.8.6
```
2. At this point, the node module for spotify should be built. To get soundbounce frontend running...
   * Put the node module you just built on node's path:
```
  $ export NODE_PATH=<your-path>/soundbounce/src/client-node-webkit/node-spotify-master/build/Release
```
   * Get the 'app.key' file from the osx distribution (I used dmg2img and then mounted the image) and put it in next to osx.html.
   * Run `nw .` in the directory with osx.html
3. I had to login and restart several times before the sound started working. After joining a room, I have to stay on the 'room list' page to have the playback continue through song changes.

Hope that helps.

**Comment by earthling** [comment]
I don't think you need to `npm install node-libspotify`, the node-spotify-master _is_ node-libspotify. And it needs to be built with nw-gyp. Make sure you have soundbounce/src/client-node-webkit/node-spotify-master/build/Release on the NODE_PATH.
I found the 'app.key' file inside a zip in: Soundbounce-v0.9-beta-osx1/Soundbounce.app/Contents/Resources/app.nw . To be honest, I gave up trying to mount the dmg/img from the command line and just used nautilus to open (i.e., mount) the converted soundbounce.img file and again to open the 'app.nw' file (I've no idea what sort of compression that uses). The 'app.key' file is in 'app.nw'.

**Comment by carnager** [comment]
ok, got the app.key file. 
as you can see i did the export, but it doesnt work:
https://fb.53280.de/laC2/

**Comment by earthling** [comment]
Did `nw-gyp configure --target=0.8.6` and `nw-gyp build --target=0.8.6` succeed in node-spotify-master? I also ran `nw-gyp install` at some point, but I think all that does is 'install' the nodejs headers it needs. I don't think it 'installs' any build artifacts.

**Comment by carnager** [comment]
make: Leaving directory '/home/carnager/soundbounce/src/client-node-webkit/node-spotify-master/build'
gyp info ok 


yes, worked fine. however nw-gyp install fails...

**Comment by earthling** [comment]
Does your build directory look like this?
```
ls build/Release  
linker.lock  metadataUpdater.js  nodespotify.node  obj.target  spotify.js
```
Also, at one point I made a misguided effort to run the server and installed a bunch of packages with npm, maybe try this: `sudo npm install spotify-web-api-node`?

**Comment by carnager** [comment]
carnager@caprica ~/soundbounce/src/client-node-webkit/node-spotify-master > ls build/Release  
obj.target/  linker.lock  metadataUpdater.js  nodespotify.node*  spotify.js

**Comment by earthling** [comment]
Sorry man, that's about all I got. If I unset NODE_PATH I get the popup warning about not being able to find libspotify, but I do *not* get any errors on the console about 'Uncaught TypeError'. I just see:
```
[8313:0106/164536:INFO:CONSOLE(343)] "Uncaught Error: Cannot find module 'spotify'", source: module.js (343)
```
Are you able to login despite the error?

**Comment by carnager** [comment]
no... libspotify is indeed not loaded, so it complains about unimplemented functions.

[12680:0107/015315:INFO:CONSOLE(80)] "Uncaught TypeError: Cannot call method 'login' of undefined", source: http://app.soundbounce.org/login.html (80)

**Comment by earthling** [comment]
I'm on Ubuntu 14.10 - you?

**Comment by carnager** [comment]
arch linux

**Comment by carnager** [comment]
I suppose the most striking thing is the segfault i get, when running the compiled node-spotify binary...

**Comment by pdaddyo** [comment]
Hi guys,  If you got this working please can you post a build or some instructions for getting it to work on Linux here?  Would love to point people this way if they're on linux.  Cheers

**Comment by carnager** [comment]
ok, my issue was something else: i forgot the dash in the linker settings. so i had "libspotify" instead of "-libspotify"

so my issue is now that it doesnt find libspotify at build time: 

arnager@caprica ~/soundbounce/src/client-node-webkit/node-spotify-master > nw-gyp build --target=0.8.6            
gyp info it worked if it ends with ok
gyp info using nw-gyp@0.12.4
gyp info using node@0.10.35 | linux | x64
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '/home/carnager/soundbounce/src/client-node-webkit/node-spotify-master/build'
  SOLINK_MODULE(target) Release/obj.target/nodespotify.node
/usr/bin/ld: cannot find -libspotify

**Comment by carnager** [comment]
OK, it was relatively easy and mainly caused by a typo, that it didn't work.
Here are the steps:

* git clone https://github.com/pdaddyo/soundbounce.git
* cd soundbounce/src/client-node-webkit/node-spotify-master/
* vim binding.gyp - change it as follows:
```
    "link_settings" : { 
      "libraries": ["-lasound -lspotify"]
    },  
```
* sudo npm install -g nw-gyp (I had to do a global install, otherwise it didn't work)
* npm install nodewebkit@0.8.6 
* nw-gyp configure --target=0.8.6 --python=python2
* nw-gyp build --target=0.8.6 --python=python2
* cd ../osx
* cp /path/to/app.key .
* nw .

enjoy!

**Comment by carnager** [comment]
a hopefully working 64bit binary:
https://dl.53280.de/soundbounce_linux.tar.bz2

extract it. 
cd to soundbounce/
run ./start.sh

**Comment by pdaddyo** [comment]
Thanks for these instructions, appreciated.

**Comment by carnager** [comment]
Arch Linux Package in AUR:

https://aur.archlinux.org/packages/soundbounce-git/

**Comment by ikt** [comment]
./start.sh 
[23617:0111/183448:INFO:CONSOLE(359)] "Uncaught Error: libspotify.so.12: cannot open shared object file: No such file or directory", source: module.js (359)

Thanks for the work! Where does one find libspotify.so.12? Using Ubuntu 14.04

**Comment by carnager** [comment]
thanks, added libspotify to dependencies for the arch package.

**Comment by ikt** [comment]
Quick question @pdaddyo is the v1 release expected to be the last release for a while? It looks like most of the work is happening on the server side of things, should I assume the v1 release will be around for a while? thanks

**Comment by pdaddyo** [comment]
I had hoped to get v1.1 out of the door by now, with items that need to be on the native client side (volume control, selecting temp files location etc), but unfortunately real life has got in the way.  Realistically, it'll be at least another month until I'm able to even look at items.  So we're looking at May at the earliest, and to be honest it'll probably be a windows-only update since I'm looking for help with the node-spotify / osx / linux bugs that users are encountering before adding features (these aren't my daily platforms).  Thanks.

**Comment by ikt** [comment]
thanks for the info :)

**Comment by pdaddyo** [comment]
Does the Chrome Extension count as Linux support? :)

**Comment by ikt** [comment]
Very cool and I'm installing it, but def want ubuntu support, may be possible with the upcoming snap packages.

https://developer.ubuntu.com/en/snappy/tutorials/node-to-snap/
